static/app/components/forms/teamSelector.tsx

```diff
@@ -0,0 +1,265 @@
+  }),
+};
+
+type Props = {
+  api: Client;
+  organization: Organization;
+  teams: Team[];
+  onChange: (value: any) => any;
+  /**
+   * Function to control whether a team should be shown in the dropdown
+   */
+  teamFilter?: (team: Team) => boolean;
+  /**
+   * Can be used to restrict teams to a certain project and allow for new teams to be add to that project
+   */
+  project?: Project;
+  /**
+   * Controls whether the value in the dropdown is a team id or team slug
+   */
+  useId?: boolean;
+  includeUnassigned?: boolean;
+} & ControlProps;
+
+type TeamActor = {
+  type: 'team';
+  id: string;
+  name: string;
+};
+
+type TeamOption = {
+  value: string | null;
+  label: React.ReactElement;
+  searchKey: string;
+  actor: TeamActor | null;
+  disabled?: boolean;
+};
+
+type State = {
+  options: TeamOption[];
+};
+
+class TeamSelector extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      options: this.teamOptions,
+    };
+  }
+
+  // TODO(ts) This type could be improved when react-select types are better.
+  selectRef = React.createRef<any>();
+
+  get teamOptions() {
+    const {teams, teamFilter, includeUnassigned, project} = this.props;
+    const filteredTeams = teamFilter ? teams.filter(teamFilter) : teams;
+    if (project) {
+      const teamsInProjectIdSet = new Set(project.teams.map(team => team.id));
+      const teamsInProject = filteredTeams.filter(team =>
+        teamsInProjectIdSet.has(team.id)
+      );
+      const teamsNotInProject = filteredTeams.filter(
+        team => !teamsInProjectIdSet.has(team.id)
+      );
+
+      return [
+        ...teamsInProject.map(this.createTeamOption),
+        ...teamsNotInProject.map(this.createTeamOutsideProjectOption),
+        ...(includeUnassigned ? [unassignedOption] : []),
+      ];
+    }
+    return [
+      ...filteredTeams.map(this.createTeamOption),
+      ...(includeUnassigned ? [unassignedOption] : []),
+    ];
+  }
+
+  createTeamOption = (team: Team): TeamOption => ({
+    value: this.props.useId ? team.id : team.slug,
+    label: <IdBadge team={team} />,
+    searchKey: `#${team.slug}`,
+    actor: {
+      type: 'team',
+      id: team.id,
+      name: team.slug,
+    },
+  });
+
+  /**
+   * Closes the select menu by blurring input if possible since that seems to be the only
+   * way to close it.
+   */
+  closeSelectMenu() {
+    if (!this.selectRef.current) {
+      return;
+    }
+
+    const select = this.selectRef.current.select;
+    const input: HTMLInputElement = select.inputRef;
+    if (input) {
+      // I don't think there's another way to close `react-select`
+      input.blur();
+    }
+  }
+
+  handleAddTeamToProject = async (team: Team) => {
+    const {api, organization, project, value} = this.props;
+    const {options} = this.state;
+
+    if (!project) {
+      this.closeSelectMenu();
+      return;
+    }
+
+    // Copy old value
+    const oldValue = value ? [...value] : {value};
```

Comment:
I'm a bit confused about this condition. If value quotes to falsy why would it be `{value}`?

Reply:
yea I moved this over from selectMembers.tsx and was too scared to change it in case this handled some bits of react-select i didn't know about. 
My current understanding of this is that to support the `multiple` react-select we have to supply an array of values. But it seems like that array of values also works for react-select dropdowns which only support one value. So doing the [...value] will satisfy both types of dropdowns. And for the `{value}` this is because `this.props.onChange` expects an object like `{value: ...}` so if value is falsy we can't do `[...value]` and instead just pass something like `{value: null}` which will be consumed properly by the `onChange` function in parent components.

Reply:
This feels pretty messy though I think im gonna clean this up so it's easier to understand

Reply:
cool, yeah sounds good